coldfront/plugins/qumulo/management/commands/check_billing_cycles.py

```diff
@@ -0,0 +1,126 @@
+                allocation=allocation,
+                allocation_attribute_type=billing_attribute,
+            ).update(value="prepaid")
+            AllocationAttribute.objects.filter(
+                allocation=allocation,
+                allocation_attribute_type=service_rate_attribute,
+            ).update(value="subscription")
+
+
+def calculate_prepaid_expiration(
+    allocation, bill_cycle, prepaid_months, prepaid_billing_start, prepaid_expiration
+) -> None:
+    logger.warn(f"Calculation prepaid expiration")
+    prepaid_exp_attribute = AllocationAttributeType.objects.get(
+        name="prepaid_expiration"
+    )
+    if bill_cycle == "prepaid" and prepaid_expiration == None:
+        prepaid_billing_start = datetime.strptime(prepaid_billing_start, "%Y-%m-%d")
+        prepaid_months = int(prepaid_months)
+        prepaid_until = datetime(
+            prepaid_billing_start.year
+            + (prepaid_billing_start.month + prepaid_months - 1) // 12,
+            (prepaid_billing_start.month + prepaid_months - 1) % 12 + 1,
+            prepaid_billing_start.day,
+        )
+        AllocationAttribute.objects.get_or_create(
+            allocation=allocation,
+            allocation_attribute_type=prepaid_exp_attribute,
+            value=prepaid_until,
+        )
+
+
+def check_allocations() -> None:
+    resource = Resource.objects.get(name="Storage2")
+    allocations = Allocation.objects.filter(status__name="Active", resources=resource)
+    billing_attribute = AllocationAttributeType.objects.get(name="billing_cycle")
+    prepaid_exp_attribute = AllocationAttributeType.objects.get(
+        name="prepaid_expiration"
+    )
+    prepaid_billing_start_attribute = AllocationAttributeType.objects.get(
+        name="prepaid_billing_date"
+    )
+    prepaid_months_attribute = AllocationAttributeType.objects.get(name="prepaid_time")
+    billing_sub_q = AllocationAttribute.objects.filter(
+        allocation=OuterRef("pk"), allocation_attribute_type=billing_attribute
+    ).values("value")[:1]
+    prepaid_exp_sub_q = AllocationAttribute.objects.filter(
+        allocation=OuterRef("pk"), allocation_attribute_type=prepaid_exp_attribute
+    ).values("value")[:1]
+    prepaid_billing_date_sub_q = AllocationAttribute.objects.filter(
+        allocation=OuterRef("pk"),
+        allocation_attribute_type=prepaid_billing_start_attribute,
+    ).values("value")[:1]
+    prepaid_months_sub_q = AllocationAttribute.objects.filter(
+        allocation=OuterRef("pk"),
+        allocation_attribute_type=prepaid_months_attribute,
+    ).values("value")[:1]
+    allocations = allocations.annotate(
+        billing_cycle=Subquery(billing_sub_q),
+        prepaid_expiration=Subquery(prepaid_exp_sub_q),
+        prepaid_billing_start=Subquery(prepaid_billing_date_sub_q),
+        prepaid_months=Subquery(prepaid_months_sub_q),
+    )
+    logger.warn(f"Checking billing_cycle in {len(allocations)} qumulo allocations")
```

Comment:
@cdeanj2 I would divide this method in two: one for gathering a resultset of allocations and the other for updating the billing cycle type and expiration dates.